File: C++/프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.cpp
#include <vector>
#include <stack>

using namespace std;
stack<pair<int, int>> st;
int solution(vector<int> ingredient) {
    int answer = 0;
    vector<int> order = {1, 2, 3, 1};
    
    for(auto& ing : ingredient) {
        if(!st.empty() && order[st.top().second + 1] == ing) {
            st.push({ing, st.top().second + 1});
        } else if (ing == 1) {
            st.push({ing, 0});
        } else {
            st.push({ing, -1});
        }
        
        if(!st.empty() && st.top().second == 3) {
            st.pop(); st.pop(); st.pop(); st.pop();
            answer++;
        }
    }
    
    return answer;
}

File: Python3/프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.py
def solution(ingredient):
    answer = 0
    stack, order = [], [1, 2, 3, 1]
    
    for ing in ingredient:
        if stack and order[(stack[-1] + 1) % 4] == ing:
            stack.append(stack[-1] + 1)
        else:
            stack.append(0 if ing == 1 else -1)
        
        if stack and stack[-1] == 3:
            for i in range(4):
                stack.pop();
            answer+= 1
    
    
    return answer